DataScience/Experimentation.py

```diff
@@ -86,23 +86,61 @@ def result_writer(command_list):
         experiment_file.write(line + "\n")
     experiment_file.flush()
     
-def run_experiment(command):
+def extract_loss_lines(process, threshold):
+    counter = 0
+    loss_lines = []
+    while True:
+        line = process.stdout.readline().strip()
+        counter += 1
+        if not process.poll() is None:
+            break
+        if line.startswith('average loss = '):
+            loss_lines.append(line)
+
+        if counter > threshold:
+            return None
+    return loss_lines
+
+def run_experiment(command, threshold=0):
+    if threshold <= 0:
+        return run_experiment_legacy(command)
+
+    try:
+        process = Popen(command.full_command.split(' '), universal_newlines=True, encoding='utf-8', stdout=PIPE, stderr=STDOUT)
+        loss_lines = extract_loss_lines(process, threshold)
+        if loss_lines is None:
+            Logger.error(f'Command {command.full_command} seems too have too many warnings and excluded from consideration')
+            process.stdout.close()
+            process.kill()
+        else:
+            process.stdout.close()
+            process.wait()
+            if len(loss_lines) == 1:
+                command.loss = float(loss_lines[0].split()[3])
+                Logger.info("Ave. Loss: {:12}Policy: {}".format(str(command.loss),command.full_command))
+            else:
+                Logger.error("Error for command {0}: {1} lines with 'average loss = '. Expected 1".format(command.full_command, len(loss_lines)))
+    except:
+        Logger.exception("Error for command {}".format(command.full_command))
+    return command
+
+def run_experiment_legacy(command):
     try:
-        results = check_output(command.full_command.split(' '), stderr=STDOUT).decode("utf-8")
+        results = check_output(command.full_command.split(' '), stderr=STDOUT, universal_newlines=True)
         loss_lines = [x for x in str(results).splitlines() if x.startswith('average loss = ')]
         if len(loss_lines) == 1:
             command.loss = float(loss_lines[0].split()[3])
-            Logger.info("Ave. Loss: {:12}Policy: {}".format(str(command.loss),command.full_command))
+            print("Ave. Loss: {:12}Policy: {}".format(str(command.loss),command.full_command))
         else:
-            Logger.error("Error for command {0}: {} lines with 'average loss = '. Expected 1".format(command.full_command, len(loss_lines)))
-    except:
-        Logger.exception("Error for command {}".format(command.full_command))
+            print("Error for command {0}: {} lines with 'average loss = '. Expected 1".format(command.full_command, len(loss_lines)))
+    except Exception as e:
+        print("Error for command {}: {}".format(command.full_command, e))
```

Comment:
Why are these logger lines changed to print?

Reply:
ouch. Thanks for noticing. Let me fix that.